src/peer_manager.rs

```diff
@@ -250,20 +261,34 @@ impl PeerMap {
     }
 }
 
+/// (whether requires resource proof, own group)
+type ExpectJoinResult = Result<(bool, (Prefix<XorName>, Vec<PublicId>)), RoutingTableError>;
+
 /// A container for information about other nodes in the network.
 ///
 /// This keeps track of which nodes we know of, which ones we have tried to connect to, which IDs
 /// we have verified, whom we are directly connected to or via a tunnel.
 pub struct PeerManager {
+    /// Token map for the connections
     connection_token_map: HashMap<u32, PublicId>,
+    /// Peer map for the connections
     peer_map: PeerMap,
     // Peers we connected to but don't know about yet
     unknown_peers: HashMap<PeerId, Instant>,
     // Peers we expect to connect to
     expected_peers: HashMap<XorName, Instant>,
     proxy_peer_id: Option<PeerId>,
+    /// Routing table
     routing_table: RoutingTable<XorName>,
+    /// Our public id
     our_public_id: PublicId,
+    /// A joining node wants to join our group
+    /// (name, time_inserted, seed_for_resource_proof)
+    node_candidate: Option<(XorName, Instant, Vec<u8>)>,
```

Comment:
there is chance that: we received the `approval` first, and frees `node_candidate` to accept next joining request, which arrives BEFORE the connection to the previous joining node got established to us.
This situation will be problematic to handle, if we merge this two member fields into one enum or struct.  :(